tests/extra_source_and_tests/extra_src/for_2D_build/anisotropic_diffusion_relaxation.hpp

```diff
@@ -0,0 +1,446 @@
+/**
+ * @file anisotropic_diffusion_relaxation.hpp
+ * @brief This the methods related on the corrections of SPH anisotropic smoothing kernel and anisotropic diffusion algorithm.
+ * @details Two template are included, one is the improved kernel matrix, and the other one is anisotropic diffusion relaxation.
+ * @author Xiaojing Tang and Xiangyu Hu
+ */
+
+#ifndef ANISOTROPIC
+#define ANISOTROPIC
+
+#include "base_general_dynamics.h"
+#include "sphinxsys.h"  
+namespace SPH
+{
+
+
+//TODO:: The class name. The datatype should be applicable to 6d.
+
+//material definition
+class AnisotropicDiffusionSolid : public Solid
+{
+  public:
+    AnisotropicDiffusionSolid(Real rho0, Real coeff)
+        : Solid(rho0), diffusion_coeff(coeff)
+    {
+        material_type_name_ = "AnisotropicDiffusionSolid";
+    };
+    virtual ~AnisotropicDiffusionSolid(){};
+
+    Real diffusion_coeff;
+    Real DiffusivityCoefficient() { return diffusion_coeff; };
+};
+ 
+//----------------------------------------------------------------------
+//	calculate correction matrix B to keep the accuracy in anisotropic case
+//----------------------------------------------------------------------
+template <typename... InteractionTypes>
+class AnisotropicLinearGradientCorrectionMatrix;
+
+template <class DataDelegationType>
+class AnisotropicLinearGradientCorrectionMatrix<DataDelegationType>
+    : public LocalDynamics, public DataDelegationType
+{
+  public:
+    template <class BaseRelationType>
+    explicit AnisotropicLinearGradientCorrectionMatrix(BaseRelationType &base_relation)
+     : LocalDynamics(base_relation.getSPHBody()), DataDelegationType(base_relation),
+       Vol_(this->particles_->template getVariableDataByName<Real>("VolumetricMeasure")),
+       B_(particles_->registerStateVariable<Matd>("LinearGradientCorrectionMatrix", IdentityMatrix<Matd>::value)),
+      pos_(this->particles_->template getVariableDataByName<Vecd>("Position"))  {}
+
+    virtual ~AnisotropicLinearGradientCorrectionMatrix(){};
+
+  protected:
+    Real *Vol_;
+    Matd *B_;
+    Vecd *pos_;
+
+};
+
+template <>
+class AnisotropicLinearGradientCorrectionMatrix<Inner<>>
```

Comment:
This exact the same as B computed in the library. I do not think that we need a new class for it.